### ISSUE TITLE
Job hangs if 67 000 files were added since last backup

### ISSUE THREAD

**Issue by d-bykov** [duplicati/duplicati#4172]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.5.104_canary and 2.0.5.1_beta
- **Operating system**: Windows 10 Pro x64 1803
- **Backend**: SFTP

## Description

During backup, the job may hang with "XXXXX files (Y.YY GB) to go" status message, which doesn't change for hours (probably won't change at all). Current action is shown as "Backup_ProcessingFiles", Current file doesn't change either.
Trying to see log in this state gives me "Failed to connect: database is locked database is locked" error. After stopping the "Duplicati service", the "Duplicati.Server.exe" remains running and fully utilizes one CPU core until I kill it manually. After that, if I start the Duplicati service, the backup job starts immediately (I set up schedule to run every 6 minutes) and then freezes with the same status.
~~The only way I found to make the job run to the end is to disable the USN, kill the "Duplicati.Server.exe" and restart Duplicati service. After that, the job runs completes successfully.
I was hoping to replace CrashPlan real-time protection using the Duplicati with a job that is scheduled to run even few minutes, but without USN enabled, Duplicati re-scans all the source folder (in my case it's about 90k files with total size of 2.75GB) taking about two minutes with relatively high CPU and disk utilization, making work uncomfortable.
With USN enabled, the job takes less than 15 seconds, not affecting the overall user experience, but it only works if no major file changes were made during the job execution.
Setting the snapshot-policy to Required and Off doesn't have any affect on the job freezing.~~

**Update: the USN seems to be irrelevant.**

## Steps to reproduce

1. Create an empty source folder
2. Create a backup job and run it
3. Put a subfolder with a lot of files to the source folder (I used the latest Linux kernel source)
4. Run the job again

- **Actual result**:
Job execution hangs, the status remains "XXXXX files (Y.YY GB) to go" for indefinitely long, with Current action = "Backup_ProcessingFiles", "Current file" field value doesn't change. 
Duplicati.Server.exe fully utilizes 1 CPU core.
- **Expected result**:
Successful job execution.

When it happens, the Home page looks like this, where the transfer speed is decreasing little by little, but the rest info doesn't change at all:

![image](https://user-images.githubusercontent.com/25761787/79600653-cc344600-80e7-11ea-89d6-d5f615662aac.png)

In the Sysinternals Process Monitor I can see, that the Duplicati.Server.exe is constantly reads the SQLite database, but there is no other file activity:

![79457569-ba23ac00-7ff0-11ea-90e7-72b3c3450203](https://user-images.githubusercontent.com/25761787/79600698-dfdfac80-80e7-11ea-9041-51367653a3fb.png)

**Comment by drwtsn32x** [comment]
I don't understand why using snapshots wouldn't change this behavior. The whole point of a snapshot is that it's guaranteed to be a static point-in-time view of the filesystem.  Making changes while a backup is running doesn't affect the snapshot.

Can you try setting --snapshot-policy to Required to make sure that it's definitely taking one?

**Comment by d-bykov** [comment]
> Can you try setting --snapshot-policy to Required to make sure that it's definitely taking one?
I wrote that I tried to set snapshot-policy to On, but actually I was setting it to Required and there was no difference in Duplicati behavior - the job hanged with both values.

**Comment by drwtsn32x** [comment]
Very interesting. Obviously it shouldn't hang regardless of snapshot setting. I'll try to reproduce this.  Have you been able to reproduce on a second machine?

**Comment by d-bykov** [comment]
> Have you been able to reproduce on a second machine?

Yes.
This is a sequence I triggered it on another machine:

1. Create an empty folder, create a subfolder inside and fill it with a lot of files (I used latest Linux kernel source)
2. Create a backup job with both snapshot-policy and usn-policy set to Required 
3. Run the job, and somewhere in the middle, after all files were counted and now are being transferred to a destination, delete the subfolder.

In my case, the Home page looks like this, where the transfer speed is decreases little by little, but the rest info doesn't change at all:

![image](https://user-images.githubusercontent.com/25761787/79457107-f1458d80-7fef-11ea-92a5-274c29bc3c2e.png)

In the Sysinternals Process Monitor I can see, that the Duplicati.Server.exe is constantly reads the SQLite database, but there is no other file activity:

![image](https://user-images.githubusercontent.com/25761787/79457569-ba23ac00-7ff0-11ea-90e7-72b3c3450203.png)

**Comment by d-bykov** [comment]
OK, today I was able to make a job hang even with the usn-policy set to Off.
It worked fine with my normal daily use case scenario, but hanged while making a synthetic test with the Linux kernel from my previous post.
I updated the issue.

**Comment by d-bykov** [comment]
I added a little hack that fixes the problem for me: https://github.com/duplicati/duplicati/compare/master...d-bykov:sqlite-hang-fix?expand=1

By no mean it's a ready-to-use solution, just a little hint on what's could be wrong.

**Comment by surlyhacker** [comment]
I have this same issue.  Job hangs forever.  Looking in process monitor shows lots of sqlite file access and no other file access.  Using process explorer to examine the thread call stack shows all sqlite stuff. Running an "analyze" on the sqlite db fixes it.  It does seem to me that having USN turned on triggers this issue in the sense that if USN is actually used then the hang occurs later on in the backup process.  If USN is not used (see issue #4175) then I believe the hang does not occur.  But I am not 100% sure.  Haven't had time to test enough.

**Comment by d-bykov** [comment]
My previous fix didn't work well, today I had an another hang.
Now I made Duplicati to run ANALYZE on every connection to local database - https://github.com/duplicati/duplicati/commit/3a49bd9ae3d7954d37808bd77536a2514b27bfba

I played with it for a couple of hours, it worked well. I will keep an eye on it.

**Comment by brummelte** [comment]
For me (2.0.5.106_canary_2020-05-11) it seems to also access a file in the temp directory a lot:

![image](https://user-images.githubusercontent.com/7269775/81862772-7e85ee80-956a-11ea-90a6-deba79645266.png)

**Comment by michael9dk** [comment]
I'm able to reproduce it in 2.0.5.1_beta (public release).
It apears to start over again (as in the above image), when it reaches EOF.

I'll try to see if I can catch it in debug-mode.

**Comment by Alex131089** [comment]
Hello, 

I encountered this bug too on 2.0.5.1_beta_2020-01-18.

Custom global options:
```
--full-result=true
--usn-policy=Auto
--snapshot-policy=Auto
--upload-verification-file=true
--use-block-cache=true
--auto-vacuum=true
--auto-vacuum-interval=1M
--log-retention=18M
--log-file=C:\Users\Alexandre\AppData\Roaming\Duplicati\verbose.log
--log-file-log-level=Verbose
--console-log-level=Verbose
```

Duplicati started without admin privileges (so USN not usable ?), hang (I thought it was because I renamed a parent folder while backuping, even though this shouldn't be an issue), I finally killed it, then restarted the backup, same hang (with a lot of read IO), killed, restarted with `usn-policy` set to `off` -> ok. Maybe just a coincidence but it worked changing this setting.

Verbose logs (btw, there's no way to set their locale to Invariant it seems, empty [`force-locale`](https://duplicati.readthedocs.io/en/latest/06-advanced-options/#force-locale) option is removed by GUI), stuck on the last line 
```
2021-01-05 13:13:09 +01 - [Information-Duplicati.Library.Main.Controller-StartingOperation]: The operation Backup has started
2021-01-05 13:13:13 +01 - [Verbose-Duplicati.Library.Snapshots.WindowsSnapshot-WindowsSnapshotCreation]: Failed to initialize windows snapshot instance
System.UnauthorizedAccessException: Tentative d'exécution d'une opération non autorisée.
   à Alphaleonis.Win32.Vss.VssBackupComponents..ctor()
   à Alphaleonis.Win32.Vss.VssImplementation.CreateVssBackupComponents()
   à Duplicati.Library.Common.IO.VssBackupComponentsHelper.GetVssBackupComponents()
   à Duplicati.Library.Snapshots.WindowsSnapshot..ctor(IEnumerable`1 sources, IDictionary`2 options)
2021-01-05 13:13:13 +01 - [Verbose-Duplicati.Library.Snapshots.WindowsSnapshot-VSSCleanupOnError]: Failed during VSS error cleanup
System.NullReferenceException: La référence d'objet n'est pas définie à une instance d'un objet.
   à Duplicati.Library.Snapshots.WindowsSnapshot.Dispose(Boolean disposing)
   à Duplicati.Library.Snapshots.SnapshotBase.Dispose()
   à Duplicati.Library.Snapshots.WindowsSnapshot..ctor(IEnumerable`1 sources, IDictionary`2 options)
[..]
2021-01-05 13:13:40 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Started:  ()
2021-01-05 13:13:56 +01 - [Information-Duplicati.Library.Main.BasicResults-BackendEvent]: Backend event: List - Completed:  (2.24 KB)
2021-01-05 13:13:56 +01 - [Verbose-Duplicati.Library.Main.Operation.Backup.MetadataPreProcess.FileEntry-AddDirectory]: Adding directory ......
[..]
2021-01-05 13:14:51 +01 - [Verbose-Duplicati.Library.Main.Operation.Backup.FileBlockProcessor.FileEntry-NewFile]: New file C:\PathToFile
```

**Comment by drwtsn32x** [comment]
@Alex131089 - both the USN and the Snapshot options require Duplicati be running with elevated/admin rights.

> I thought it was because I renamed a parent folder while backuping, even though this shouldn't be an issue

That almost certainly would cause a problem unless you were using snapshots. But even then, you'd have to reconfigure your backup job afterward since it's one of the 'top level' folders you picked for backup. (If I'm understanding you correctly.)

**Comment by ts678** [comment]
[2.0.5.111_canary_2020-09-26](https://github.com/duplicati/duplicati/releases/tag/v2.0.5.111-2.0.5.111_canary_2020-09-26) (so in [2.0.6.3_beta_2021-06-17](https://github.com/duplicati/duplicati/releases/tag/v2.0.6.3-2.0.6.3_beta_2021-06-17) and beyond)

> Improved performance of USN, thanks @dgehri

Is anybody still seeing this issue? I think I saw it in a different scenario, maybe converting a backup to USN, but I'm not certain.
I remember seeing it doing way too much SQL, and I think I decided that it was doing little SQL operations way too frequently.

[Untraceable long operation (query?) during backup](https://forum.duplicati.com/t/untraceable-long-operation-query-during-backup/9007) and read the end. Both methods may have slow cases. Asking for feedback.
I think the latest complaint here was on 2.0.5.1 which predates the change. Lacking feedback, I'll assume the new way is better.
